integration/distinct_compat_test.go

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/FerretDB/FerretDB/integration/setup"
+	"github.com/FerretDB/FerretDB/integration/shareddata"
+)
+
+// distinctCompatTestCase describes count compatibility test case.
+type distinctCompatTestCase struct {
+	field         string                   // required
+	skipForTigris string                   // optional
+	filter        bson.D                   // required
+	resultType    compatTestCaseResultType // defaults to nonEmptyResult
+}
+
+func testDistinctCompat(t *testing.T, testCases map[string]distinctCompatTestCase) {
+	t.Helper()
+
+	// Use shared setup because distinct queries can't modify data.
+	// TODO Use read-only user. https://github.com/FerretDB/FerretDB/issues/1025
+	s := setup.SetupCompatWithOpts(t, &setup.SetupCompatOpts{
+		Providers:                shareddata.AllProviders(),
+		AddNonExistentCollection: true,
+	})
+	ctx, targetCollections, compatCollections := s.Ctx, s.TargetCollections, s.CompatCollections
+
+	for name, tc := range testCases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+
+			if tc.skipForTigris != "" {
+				setup.SkipForTigrisWithReason(t, tc.skipForTigris)
+			}
+
+			t.Parallel()
+
+			filter := tc.filter
+			require.NotNil(t, filter, "filter should be set")
+
+			var nonEmptyResults bool
+			for i := range targetCollections {
+				targetCollection := targetCollections[i]
+				compatCollection := compatCollections[i]
+				t.Run(targetCollection.Name(), func(t *testing.T) {
+					t.Helper()
+
+					targetRes, targetErr := targetCollection.Distinct(ctx, tc.field, tc.filter)
+					compatRes, compatErr := compatCollection.Distinct(ctx, tc.field, tc.filter)
+
+					if targetErr != nil {
+						t.Logf("Target error: %v", targetErr)
+						targetErr = UnsetRaw(t, targetErr)
+						compatErr = UnsetRaw(t, compatErr)
+						assert.Equal(t, compatErr, targetErr)
+						return
+					}
+					require.NoError(t, compatErr, "compat error; target returned no error")
+
+					t.Logf("Compat (expected) result: %v", compatRes)
+					t.Logf("Target (actual)   result: %v", targetRes)
+
+					require.Equal(t, len(compatRes), len(targetRes))
+
+					// If compat result is an empty array, the target results must be empty array too
+					// (not nil or something else).
+					if len(compatRes) == 0 {
+						assert.Equal(t, compatRes, targetRes)
+					}
+
+					// We can't check the exact data types because they might be different.
+					// For example, if targetRes is [float64(0), int32(1)] and compatRes is [int64(0), int64(1)],
+					// we consider them equal. If different documents use different types to store the same value
+					// in the same field, it's hard to predict what type will be returned by distinct.
```

Comment:
Do we have any plans for future to investigate how MongoDB handles every case, or that's too much unnecessary effort?

Reply:
That looks like an unnecessary effort to me, as it's such a tiny detail. I suggest keeping it as is and only implementing it if we have a user request. Also, maybe in some cases distinct can be moved to pushdown side, and this is another reason why I don't want to invest into it now. 

In principle, we can document it as a known difference and write a diff test for it. But again, that looks unnecessary to me.